SonarQube.Client/Models/ServerSentEvents/ServerContract/SqSSEStreamReader.cs

```diff
@@ -43,31 +43,46 @@ internal interface ISqSSEStreamReader : IDisposable
     /// </summary>
     internal sealed class SqSSEStreamReader : ISqSSEStreamReader
     {
-        private readonly StreamReader networkStreamReader;
-        private readonly CancellationToken cancellationToken;
+        private readonly string projectKey;
+        private readonly ISSEConnectionFactory sseConnectionFactory;
+        private const int MaxReconnectionsCount = 5; // possibly needs to be configurable
+        private readonly CancellationToken disposalToken;
         private readonly ISqServerSentEventParser sqServerSentEventParser;
+        private StreamReader currentNetworkStreamReader;
+        private int reconnectionsCount = -1;
 
-        public SqSSEStreamReader(StreamReader networkStreamReader, CancellationToken cancellationToken)
-            : this(networkStreamReader, cancellationToken, new SqServerSentEventParser())
+        public SqSSEStreamReader(string projectKey, ISSEConnectionFactory sseConnectionFactory, CancellationToken disposalToken)
+            : this(projectKey, sseConnectionFactory, disposalToken, new SqServerSentEventParser())
         {
         }
 
-        internal SqSSEStreamReader(StreamReader networkStreamReader,
-            CancellationToken cancellationToken,
+        internal SqSSEStreamReader(
+            string projectKey,
+            ISSEConnectionFactory sseConnectionFactory,
+            CancellationToken disposalToken,
             ISqServerSentEventParser sqServerSentEventParser)
         {
-            this.networkStreamReader = networkStreamReader;
-            this.cancellationToken = cancellationToken;
+            this.projectKey = projectKey;
+            this.sseConnectionFactory = sseConnectionFactory;
+            this.disposalToken = disposalToken;
             this.sqServerSentEventParser = sqServerSentEventParser;
         }
 
         public async Task<ISqServerEvent> ReadAsync()
         {
             var eventLines = new List<string>();
 
-            while (!networkStreamReader.EndOfStream && !cancellationToken.IsCancellationRequested)
+            while (!disposalToken.IsCancellationRequested
+                   && await CheckCurrentStreamReaderAndReconnect() 
+                   && !currentNetworkStreamReader.EndOfStream /*I recently discovered that this tries to read the underlying stream synchronously*/)
```

Comment:
oh, and one more thing - .EndOfStream can also throw, so we definitely need to get rid of it